src/core.js

```diff
@@ -118,17 +118,16 @@ var alignWithDOM = function(nodeName, key, statics) {
     return;
   }
 
-  var existingNode = getChild(currentParent, key);
   var matchingNode;
 
+  if (key) {
+    matchingNode = getChild(currentParent, key);
+  }
+
   // Check to see if the node has moved within the parent or if a new one
   // should be created
-  if (existingNode) {
-    if (process.env.NODE_ENV !== 'production') {
-      assertKeyedTagMatches(getData(existingNode).nodeName, nodeName, key);
-    }
-
-    matchingNode = existingNode;
+  if (matchingNode && process.env.NODE_ENV !== 'production') {
+    assertKeyedTagMatches(getData(matchingNode).nodeName, nodeName, key);
   } else {
     matchingNode = createNode(
```

Comment:
This logic will always cause a new node to be created. The whole `if` statement above it will be stripped because it will become `if (matchingNode && false)`.